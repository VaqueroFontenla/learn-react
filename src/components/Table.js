
import React, { Component } from 'react';
import { sortBy } from 'lodash';

//Components
import Card from './Card';
import Sort from './Sort';

import PropTypes from 'prop-types';



const SORTS = {
    NONE: list => list,
    TITLE: list => sortBy(list, 'title'),
    AUTHOR: list => sortBy(list, 'author'),
    COMMENTS: list => sortBy(list, 'num_comments').reverse(),
    POINTS: list => sortBy(list, 'points').reverse(),
};

class Table extends Component {
    constructor(props) {
        super(props);

        this.state = {
            sortKey: 'NONE',
            isSortReverse: false,
        };

        this.onSort = this.onSort.bind(this);
    }

    onSort(sortKey) {
        const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse;
        this.setState({ sortKey, isSortReverse });
    }


    render() {

        const {
            sortKey,
            isSortReverse,
        } = this.state


        const {
            list,
            toLink,
            onDismiss
        } = this.props

        const sortedList = SORTS[sortKey](list);
        const reverseSortedList = isSortReverse
            ? sortedList.reverse()
            : sortedList;

        return (
            <div className="table">
                <div className="table-header">
                    <div className="dropdown">
                        <button className="btn btn-secondary dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                            Ordenar por
                        </button>
                        <div className="dropdown-menu" aria-labelledby="dropdownMenuButton">
                            <span className="dropdown-item">
                                <Sort
                                    sortKey={'TITLE'}
                                    onSort={this.onSort}
                                    activeSortKey={sortKey}
                                >
                                    TÃ­tulo
                                </Sort>
                            </span>
                            <span className="dropdown-item">
                                <Sort
                                    sortKey={'AUTHOR'}
                                    onSort={this.onSort}
                                    activeSortKey={sortKey}
                                >
                                    Autor
                                </Sort>
                            </span>
                            <span className="dropdown-item">
                                <Sort
                                    sortKey={'COMMENTS'}
                                    onSort={this.onSort}
                                    activeSortKey={sortKey}
                                >
                                    Comentarios
                                </Sort>
                            </span>
                            <span className="dropdown-item">
                                <Sort
                                    sortKey={'POINTS'}
                                    onSort={this.onSort}
                                    activeSortKey={sortKey}
                                >
                                    Puntos
                                </Sort>
                            </span>

                        </div>
                    </div>
                </div>
                <div className="container-fluid">
                    <div className="row">
                        {reverseSortedList.map(item =>
                            <Card
                                toLink={toLink}
                                onDismiss={onDismiss}
                                item={item} />
                        )}
                    </div>
                </div>
            </div>

        );
    }
}



Table.propTypes = {
    list: PropTypes.array.isRequired,
};

export default Table;
