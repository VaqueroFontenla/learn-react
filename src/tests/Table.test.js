import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Table from '../components/Table';

describe('Table', () => {

    const props = {
        list: [
            { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
            { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
        ],
        sortKey: 'TITLE',
        isSortReverse: false,
    };

    // Jest
    it('renders', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Table {...props} />, div);
    });

    test('snapshots', () => {
        const component = renderer.create(
            <Table {...props} />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    //Enzyme

    // it('shows two items in list', () => {
    //     const element = shallow(
    //         <Table {...props} />
    //     );

    //     expect(element.find('.table-row').length).toBe(2);
    // });

});