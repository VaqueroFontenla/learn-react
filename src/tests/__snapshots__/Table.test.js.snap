// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshots 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      &gt;
      Search
      <input
        onChange={[Function]}
        type="text"
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      More
    </button>
  </div>
</div>
`;

exports[`Button snapshots 1`] = `
<button
  className=""
  type="button"
>
  Give Me More
</button>
`;

exports[`Search snapshots 1`] = `
<form>
  &gt;
  Search
  <input
    type="text"
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table snapshots 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "largeColumn": Object {
            "width": "40%",
          },
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "midColumn": Object {
            "width": "30%",
          },
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "largeColumn": Object {
            "width": "40%",
          },
        }
      }
    >
      <a>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "midColumn": Object {
            "width": "30%",
          },
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "smallColumn": Object {
            "width": "10%",
          },
        }
      }
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
